gson/src/main/resources/META-INF/proguard/gson.pro

```diff
@@ -44,15 +46,25 @@
 
 # Keep no-args constructor of classes which can be used with @JsonAdapter
 # By default their no-args constructor is invoked to create an adapter instance
--keep class * extends com.google.gson.TypeAdapter {
+-keepclassmembers class * extends com.google.gson.TypeAdapter {
   <init>();
 }
--keep class * implements com.google.gson.TypeAdapterFactory {
+-keepclassmembers class * implements com.google.gson.TypeAdapterFactory {
   <init>();
 }
--keep class * implements com.google.gson.JsonSerializer {
+-keepclassmembers class * implements com.google.gson.JsonSerializer {
   <init>();
 }
--keep class * implements com.google.gson.JsonDeserializer {
+-keepclassmembers class * implements com.google.gson.JsonDeserializer {
   <init>();
 }
+
+# If a class still exists after shrinking, and has fields annotated with @SerializedName,
+# keep the constructors
+# Note: This behavior is not officially documented somewhere; based on https://issuetracker.google.com/issues/150189783#comment11
+# and discussion in https://github.com/google/gson/pull/2397
+-if class *
```

Comment:
This seems to work as expected for R8, but it is still to me a bit unclear how `-if` behaves / is supposed to behave. See also open questsions in https://github.com/google/gson/pull/2397#issuecomment-1575236086.

Reply:
`-if` is documented on https://www.guardsquare.com/manual/configuration/usage. Just having `<init>()` (without `. . .`) here is fine. If there is no default constructor, then the unsafe instantiation will kick in. But here is actually a little semantic gotcha, as keeping the default constructor is an indirect way of telling R8 that the class is instantiated.

Reply:
The `-if class *` will match all classes which are in the _residual program_, so not all classes in the input program. This means that only if a class is otherwise used by the program and it has `<init>()` and fields annotated by `@com.google.gson.annotations.SerializedName` then both the `<init>()` and the annotated fields will be kept.

Suggesting to always have a default constructor makes sense (I saw the change with the tests and improved error message). The fallback to using Unsafe will then only be needed if using GSON with libraries the developer does not have control over. That could be solved by byte code rewriting injecting the default constructor, but I guess that is out of scope of GSON.

Reply:
Thanks for the clarification! I have adjusted the comment in `gson.pro` to link to your explanation now.